src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/RouteHandlerAnalyzer.cs

```diff
@@ -36,7 +40,23 @@ public override void Initialize(AnalysisContext context)
             var wellKnownTypes = WellKnownTypes.GetOrCreate(compilation);
             var routeUsageCache = RouteUsageCache.GetOrCreate(compilation);
 
-            context.RegisterOperationAction(context =>
+            // Pool and reuse the same list for each block.
+            var concurrentQueue = new ConcurrentQueue<List<MapOperation>>();
+            context.RegisterOperationBlockStartAction(context =>
+            {
+                if (!concurrentQueue.TryDequeue(out var mapOperations))
+                {
+                    mapOperations = new List<MapOperation>();
+                }
+
+                context.RegisterOperationAction(c => DoOperationAnalysis(c, mapOperations), OperationKind.Invocation);
+                context.RegisterOperationBlockEndAction(c => DetectAmbiguousRoutes(c, mapOperations));
+
+                mapOperations.Clear();
+                concurrentQueue.Enqueue(mapOperations);
+            });
+
+            void DoOperationAnalysis(OperationAnalysisContext context, List<MapOperation> mapOperations)
```

Comment:
Having said the above, it is probably unlikely we make the change of invoking Operation action callbacks within a single operation block in parallel as I think that would break many shipped analyzers. We already achieve parallelism by invoking callbacks for different operation blocks in parallel.